infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/LogicalGroupSnapshotReader.java

```diff
@@ -1,21 +1,177 @@
 package org.corfudb.infrastructure.logreplication.replication.send.logreader;
 
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
 import org.corfudb.infrastructure.logreplication.infrastructure.LogReplicationContext;
+import org.corfudb.infrastructure.logreplication.config.LogReplicationLogicalGroupConfig;
+import org.corfudb.protocols.logprotocol.OpaqueEntry;
+import org.corfudb.protocols.logprotocol.SMREntry;
 import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.LogReplication;
 import org.corfudb.runtime.LogReplication.LogReplicationSession;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.util.serializer.Serializers;
+
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static com.google.protobuf.UnsafeByteOperations.unsafeWrap;
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolLogReplication.generatePayload;
+import static org.corfudb.protocols.service.CorfuProtocolLogReplication.getLrEntryMsg;
 
 /**
  * Snapshot reader implementation for Logical Grouping Replication Model.
  */
+@Slf4j
 public class LogicalGroupSnapshotReader extends BaseSnapshotReader {
 
+    /**
+     * Flattened data structure of the field {@link LogReplicationLogicalGroupConfig#getRemovedGroupToStreams()} in
+     * LogReplicationLogicalGroupConfig.
+     */
+    private Set<String> streamsToClear;
+
+    private boolean isReadingStreamToClear;
+
+    private static final String CLEAR_SMR_METHOD = "clear";
+
+    private static final SMREntry CLEAR_ENTRY = new SMREntry(CLEAR_SMR_METHOD, new Array[0], Serializers.PRIMITIVE);
```

Comment:
We do not need this mechanism of adding a special clear entry for streams which were removed from a group.  'clearLocalStreams()' deletes the contents of all such streams.  LogReplicationListener will also not find anything in the diff so they will get removed.

Let me know if I am missing something.

Reply:
I think by computing the diff the merge-listener could be able to remove the LR table entries from merge table, even without this CLEAR_ENTRY. but if I understand correctly, for the LR table itself, on Sink side the listener will not be responsible for writing/updating to it. That's why we will have to clear the LR table by the snapshot writer using this CLEAR_ENTRY

Reply:
clearLocalStreams() on the snapshot writer clears all replicated tables for which no data was received.

Reply:
yes for FULL_TABLE we can rely on that, since Sink side knows the streams to replicate by checking is_federated. The limitation of LOGICAL_GROUP case is only Source side knows the "correct groups", so Sink side could only honor whatever Source sends

Reply:
But the set of streams to replicate gets constructed on the Sink for this model as well, using the 'group' options.  Any table for which data is not received, gets cleared.

Reply:
Discussed offline and we agreed that relying on writes to clear streams will work and simplify some workflows